server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java

```diff
@@ -84,6 +85,60 @@ public long readLong() throws IOException {
         }
     }
 
+    @Override
+    public String readString() throws IOException {
+        final int chars = readArraySize();
+        if (slice.hasArray()) {
+            // attempt reading bytes directly into a string to minimize copying
+            final byte[] bytes = slice.array();
+            final int start = slice.position() + slice.arrayOffset();
+            final int limit = slice.limit() + slice.arrayOffset();
+
+            if (limit - start < chars) {
+                return doReadString(chars); // not enough bytes to read at once
+            }
+
+            // calculate number of bytes matching the requested chars
+            int pos = start;
```

Comment:
This is interesting - I wonder what does it mean wrt `StreamOutput#writeString` (and the [related method I'm adding](https://github.com/elastic/elasticsearch/pull/104680/files#diff-0d24047955e824f5340c51cc91a095c8d3559ed404534c084afb7171e57340dbR462) - which of course needs to be aligned with how we actually write. 

It seems like we are doing it wrong, right? Should we fix it?

Reply:
I was wondering about writeString, unfortunately there's no easy fix with regards to backwards compatibility :/
We would have to fix the old readString to properly support 4-byte sequences for surrogates and only 2 (?) major versions later we could start writing these if we're sure they are understood on the other side :/

At this point it doesn't even seem to be worth starting the effort :(

Reply:
I think the key is enforcing ASCII whenever possible before even writing.
That allows to get rid of a lot of complicated checks when reading for these cases, such as supporting multiple length characters or even surrogate pairs 🤯

Reply:
We can always change the format, it is versioned. But I suggest handling that in a followup.

Reply:
We should definitely document the deviation from UTF-8, but I suspect we might not want to fix it in the wire protocol. Java itself doesn't really handle the top Unicode plane very well, it's surrogates all the way down, e.g.

```java
assertThat("💩", hasLength(2));
```

If we were to encode top-plane codepoints using the "correct" 4-byte UTF-8 sequences I expect we'd have to introduce quite some extra complexity (e.g. not being able to use `String#length` to count the characters we're sending) and yet I don't see an obvious benefit over what we're doing today. We control both sender and receiver, as long as they agree we can send this stuff however we want.

Reply:
> yet I don't see an obvious benefit over what we're doing today

You could build a much faster reader if you had the byte length :) no need for going through the bytes twice. For ASCII encoded strings (which are probably the vast majority) you could just `new String(` on a zero-copy slice out of buffers in most cases (rare to hit Netty buffer boundaries).
The extra cost of writing for us is mostly an implementation detail anyway. If we write a 4 byte length prefix before each string we could do so without iterating the string twice on write or so :)

Reply:
> You could build a much faster reader if you had the byte length :)

Yeah sorry I meant I don't see the benefit of fixing the deviation with UTF-8 w.r.t. surrogate pairs. There's clearly better ways to send strings over the wire. I mean how many of the strings we transmit are things like index names or field names that are already known to both sides, if we leveraged that we could probably avoid instantiating _anything_ most of the time.

Reply:
Is that going to apply to index/field names that are somewhat dynamic? Or is it just for statically-known strings like the name for a `NamedWriteable`?

Reply:
Just names of NamedWritables for now.